driver-sync/src/main/com/mongodb/client/internal/MongoClientDelegate.java

```diff
@@ -216,36 +216,29 @@ ReadBinding getReadBinding(final ReadPreference readPreference, final ReadConcer
 
         ReadWriteBinding getReadWriteBinding(final ReadPreference readPreference, final ReadConcern readConcern,
                                              @Nullable final ClientSession session, final boolean ownsSession) {
-            ReadWriteBinding readWriteBinding;
-            if (session != null && session.hasActiveTransaction()
-                    && cluster.getDescription().getType() == ClusterType.SHARDED) {
-                if (session.getPinnedMongosAddress() == null) {
-                    Server server = getCluster().selectServer(
-                            new ReadPreferenceServerSelector(getReadPreferenceForBinding(readPreference, session)));
-                    readWriteBinding = new SingleServerBinding(cluster, server.getDescription().getAddress(),
-                            getReadPreferenceForBinding(readPreference, session));
-                    session.setPinnedMongosAddress(server.getDescription().getAddress());
-                } else {
-                    readWriteBinding = new SingleServerBinding(cluster, session.getPinnedMongosAddress(),
-                            getReadPreferenceForBinding(readPreference, session));
-                }
-            } else {
-                readWriteBinding = new ClusterBinding(cluster, getReadPreferenceForBinding(readPreference, session),
-                        readConcern);
-            }
+            ReadWriteBinding readWriteBinding = new ClusterBinding(cluster, getReadPreferenceForBinding(readPreference, session),
+                    readConcern);
             if (session != null) {
                 readWriteBinding = new ClientSessionBinding(session, ownsSession, readWriteBinding);
             }
             return readWriteBinding;
         }
 
         private void labelException(final @Nullable ClientSession session, final MongoException e) {
-            if ((e instanceof MongoSocketException || e instanceof MongoTimeoutException)
-                    && session != null && session.hasActiveTransaction() && !e.hasErrorLabel(UNKNOWN_TRANSACTION_COMMIT_RESULT_LABEL)) {
+            if (session != null && session.hasActiveTransaction()
+                    && (e instanceof MongoSocketException || e instanceof MongoTimeoutException
+                    || (e instanceof MongoQueryException && e.getCode() == 91))
+                    && !e.hasErrorLabel(UNKNOWN_TRANSACTION_COMMIT_RESULT_LABEL)) {
                 e.addLabel(TRANSIENT_TRANSACTION_ERROR_LABEL);
             }
         }
 
+        private void unpinServerAddressOnTransientTransactionError(final @Nullable ClientSession session, final MongoException e) {
+            if (session != null && session.getPinnedServerAddress() != null && e.hasErrorLabel(TRANSIENT_TRANSACTION_ERROR_LABEL)) {
```

Comment:
Nit: Simplify conditional by removing null check on getPinnedServerAddress

Reply:
I have removed the null check.